src/executor/stack/mod.rs

```diff
@@ -88,59 +95,71 @@ pub struct PrecompileOutput {
 	pub logs: Vec<Log>,
 }
 
-/// Precompiles function signature. Expected input arguments are:
-///  * Address
-///  * Input
-///  * Context
-///  * State
-///  * Is static
-type PrecompileFn<S> = fn(
-	H160,
-	&[u8],
-	Option<u64>,
-	&Context,
-	&mut S,
-	bool,
-) -> Option<Result<PrecompileOutput, ExitError>>;
+/// Precompiles trait which allows for the execution of a precompile.
+pub trait Precompiles<S> {
+	/// Runs the precompile at the given address with input and context.
+	fn run(
+		&self,
+		address: H160,
+		input: &[u8],
+		gas_limit: Option<u64>,
+		context: &Context,
+		state: &mut S,
+		is_static: bool,
+	) -> Option<Result<PrecompileOutput, ExitError>>;
+
+	/// Returns the set of precompile addresses.
+	fn addresses(&self) -> &[H160];
+}
+
+#[derive(Default)]
+pub struct NoPrecompile {
+	addresses: Vec<H160>,
+}
+
+impl<S> Precompiles<S> for NoPrecompile {
+	fn run(
+		&self,
+		_address: H160,
+		_input: &[u8],
+		_gas_limit: Option<u64>,
+		_context: &Context,
+		_state: &mut S,
+		_is_static: bool,
+	) -> Option<Result<PrecompileOutput, ExitError>> {
+		None
+	}
+
+	fn addresses(&self) -> &[H160] {
+		&self.addresses
+	}
+}
 
 /// Stack-based executor.
-pub struct StackExecutor<'config, S> {
+pub struct StackExecutor<'config, S: StackState<'config>, P: Precompiles<S>> {
```

Comment:
What's the rationale for this change? One of the drawback of this is that it'll be potentially difficult for users to get a single type for `StackExecutor` suitable for different configurations of precompiles. Same reason why `Config` is not a generic parameter.

Reply:
That might've been an overzealous mistake there. I'll double check.